examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1387 @@
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/solver_control.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/sparsity_pattern.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/fe/mapping_q1.h>
+#include <deal.II/lac/sparse_direct.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <fstream>
+#include <vector>
+#include <deal.II/base/function_signed_distance.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/meshworker/mesh_loop.h>
+#include <deal.II/meshworker/scratch_data.h>
+#include <deal.II/non_matching/fe_immersed_values.h>
+#include <deal.II/non_matching/fe_values.h>
+#include <deal.II/non_matching/mesh_classifier.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_refinement.h>
+
+#ifdef USE_TRILINOS
+#  include <deal.II/lac/sparsity_tools.h>
+#  include <deal.II/lac/trilinos_vector.h>
+#  include <deal.II/lac/trilinos_block_sparse_matrix.h>
+#  include <deal.II/lac/trilinos_parallel_block_vector.h>
+#  include <deal.II/lac/trilinos_precondition.h>
+#  include <deal.II/lac/trilinos_solver.h>
+#  include <deal.II/lac/trilinos_sparse_matrix.h>
+#endif
+
+using namespace dealii;
+#ifdef USE_TRILINOS
+using VectorType = TrilinosWrappers::MPI::Vector;
+using MatrixType = TrilinosWrappers::SparseMatrix;
+#else
+using VectorType = Vector<double>;
+using MatrixType = SparseMatrix<double>;
+#endif
+
+// We will grant to some cells empty finite element spaces FE::Nothing as done
+// in Step 85,
+//  but this time active DOFS will be only assigned to cell which are
+//  intersected by the surface approximation.
+enum ActiveFEIndex
+{
+  lagrange = 0,
+  nothing  = 1
+};
+
+namespace Step86
+{
+  // using namespace dealii;
+  // @sect2{Exact solution}
+  // This class  defines the exact solution on the exact surface which is the
+  // unit sphere.
+  template <int dim>
+  class AnalyticalSolution : public Function<dim>
+  {
+  public:
+    double value(const Point<dim>  &point,
+                 const unsigned int component = 0) const override;
+
+    Tensor<1, dim> gradG(const Point<dim>  &point,
+                         const unsigned int component = 0) const;
+  };
+  // If the point of interest does not belong to the unit sphere,
+  // e.g. it is a quadrature point of a surface approximation,
+  // then the point is first projected onto the unit sphere before the
+  // evaluation.
+  template <int dim>
+  double AnalyticalSolution<dim>::value(const Point<dim>  &point,
+                                        const unsigned int component) const
+  {
+    AssertIndexRange(component, this->n_components);
+    (void)component;
+    double     distance = std::sqrt(point[0] * point[0] + point[1] * point[1] +
```

Comment:
mark any variable const if you can:
```suggestion
    const double     distance = std::sqrt(point[0] * point[0] + point[1] * point[1] +
```
Can't you just use ``point.norm()`` instead?